snapcraft/tests/unit/project_loader/test_config.py

```diff
@@ -909,6 +858,67 @@ def load_effect(*args, **kwargs):
         self.mock_load_part.assert_has_calls([call1, call2], any_order=True)
 
 
+class ValidVersionTestCase(YamlTestCase):
+    scenarios = [
+        (version, dict(version=version)) for
+        version in ['buttered-popcorn', '1.2.3', 'v12.4:1:2~', 'HeLlo']
+    ]
+
+    def test_valid_yaml_invalid_version(self):
+        fake_logger = fixtures.FakeLogger(level=logging.ERROR)
+        self.useFixture(fake_logger)
+
+        self.make_snapcraft_yaml(dedent("""
+            name: test
+            version: {!r}
+            summary: test
+            description: test
+            confinement: strict
+            grade: stable
+            parts:
+              part1:
+                plugin: nil
+        """).format(self.version))
+        _config.Config()
+
+
+class InvalidVersionTestCase(YamlTestCase):
+    scenarios = [
+        (version, dict(version=version)) for
+        version in ['*', '', '-bad', 'still-bad-', ':foo']
```

Comment:
Mind adding instances of the remaining special cases as well? That would be `.foo`, `+foo`, `~foo`, `_foo`, `bar.`, `bar:`, `bar-`, `bar_`.

Reply:
Sure, done.